lib/cudnn/upsampling.jl

```diff
@@ -0,0 +1,142 @@
+# An implementation for GPU based bilinear upsampling including its gradient
+# The code is a translation from the following files:
+# https://github.com/pytorch/pytorch/blob/master/caffe2/operators/upsample_op.cu
+# https://github.com/pytorch/pytorch/blob/master/caffe2/core/common_gpu.h
+
+# Forward and backward pass have been tested to produce the same output
+# as pytorch - it works modulo bit noise.
+
+const CUDA_NUM_THREADS = 128
+const MAXIMUM_NUM_BLOCKS = 4096
+
+@inline function GET_BLOCKS(N::Integer)
```

Comment:
Why hard-code it like that? CUDA_NUM_THREADS=128 seems awfully low for modern GPUs. Can you compare performance to using the occupancy API?

Reply:
Yes that bothered me already when writing it. It's like that because I am a GPU programming noob and this is a port from legacy code. I am not familiar with the occupancy API. Where can I find tutorials? I couldn't find any quickly.
Pytorch does sth like this nowadays
`num_threads = std::min(at::cuda::getCurrentDeviceProperties()->maxThreadsPerBlock, 1024);` [here](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cuda/UpSampleBilinear2d.cu).
How can I get the maximum number of threads per block? The `const MAXIMUM_NUM_BLOCKS` also bothers me.

Reply:
Calculating the max like that, by looking at the properties, is not guaranteed to be valid. The occupancy API can be used like this: https://github.com/JuliaGPU/CUDA.jl/blob/35150564fac7848cdf4fc3b7e74d3cca201eb442/src/indexing.jl#L32-L36

Reply:
That was fast! Thanks, I'll look into it :)

Reply:
You asked for a performance comparison; I didn't do a through comparison but noticed that the thread count tends to be  higher now. In some cases this leads to big, only partially filled blocks being launched, causing some slowdown. E.g. when I need to process 4096 pixels I get 896 threads: 4 full and a half-full one (instead of e.g. 32 full blocks with 128 threads). So I think I should artificially lower the thread count or so. But probably it's not a big issue overall.

Reply:
Yes, I've noticed something similar: https://github.com/JuliaGPU/CUDA.jl/blob/61c7879a5c1d42719d89b65cd3007061671df6c5/src/gpuarrays.jl#L15-L26. Don't have a good API answer for this yet.